### PR TITLE
feature: handle client error and repair

### LINKED ISSUE #68: [bug] dead loop when downloading files caused by that readMd5 is inconsistent with expectedMd5

### LINKED ISSUE #311: dfget卡死问题/ dfget take too long to restart
---
name: dfget卡死现象
about: dfget卡死现象原因是什么，是否是bug还是某种特定情况下会出现

---

## Question
目前在测试dragonfly，在kubernate环境中，以静态pod形式部署了两台supernode，以daemonset形式部署了客户端，dfdaemon --registry参数指定到harbor仓库。
客户端版本来自于一下官方镜像：
http://dragonfly-os.oss-cn-beijing.aliyuncs.com/df-client_0.2.0_linux_amd64.tar.gz 
在客户端运行docker pull，某些镜像下载正常，某些镜像会出现卡死现象，卡死时甚至过了进20分钟才超时。
看dfclient.log，接近20分钟一直在试图下载直到 download timeout(1149s)，:
```
[2019-01-07 14:06:32,580] ERROR sign:345064-1546868841.351 lineno:332 : piece range:20971520-25165823 error,realMd5:37a1a01672bdcf6a861615e0b73bfc03,expectedMd5:81acd03c81a67dfd115fa8900313230f,dstIp:10.16.33.2,total:521
[2019-01-07 14:06:32,588] ERROR sign:345064-1546868841.351 lineno:332 : piece range:16777216-20971519 error,realMd5:37a1a01672bdcf6a861615e0b73bfc03,expectedMd5:46989902ecda3b694a45b14f32de9ee9,dstIp:10.16.33.2,total:521
[2019-01-07 14:06:32,596] ERROR sign:345064-1546868841.351 lineno:332 : piece range:8388608-12582911 error,realMd5:37a1a01672bdcf6a861615e0b73bfc03,expectedMd5:8fb7dcfad74537126dbe1da24c3eaf48,dstIp:10.16.33.2,total:521
[2019-01-07 14:06:32,598] WARNING sign:345064-1546868841.351 lineno:167 : has not available pieceTask,maybe resource lack
[2019-01-07 14:06:32,602] ERROR sign:345064-1546868841.351 lineno:332 : piece range:12582912-16777215 error,realMd5:37a1a01672bdcf6a861615e0b73bfc03,expectedMd5:83526ce1278678e5f6688c69a78690cc,dstIp:10.16.33.2,total:521
[2019-01-07 14:06:32,604] WARNING sign:345064-1546868841.351 lineno:167 : has not available pieceTask,maybe resource lack
[2019-01-07 14:06:32,608] WARNING sign:345064-1546868841.351 lineno:167 : has not available pieceTask,maybe resource lack
[2019-01-07 14:06:32,617] ERROR sign:345064-1546868841.351 lineno:332 : piece range:0-4194303 error,realMd5:37a1a01672bdcf6a861615e0b73bfc03,expectedMd5:1d0fa006a74c02934295284d5421b09d,dstIp:10.16.33.2,total:521
[2019-01-07 14:06:32,624] INFO sign:345064-1546868841.351 lineno:110 : pull piece task result:{u'msg': u'piece resource lack', u'code': 602} and sleep 1.566 ...
[2019-01-07 14:06:32,777] ERROR sign:345064-1546868841.351 lineno:79 : download timeout(1149s)
[2019-01-07 14:06:32,779] INFO sign:345064-1546868841.351 lineno:60 : local http result:success for path:/client/ and cost:0.002
[2019-01-07 14:06:32,779] INFO sign:345064-1546868841.351 lineno:94 : |c676b9e09f098a37d06e5fb55a2d42b8abf735f00a1cb5408c7072d36455720c|https://10.250.250.16/v2/devops/jry-java/blobs/sha256:0ffa5ac9f3c5c761212037c66effa9c800973aca2f0583dcd24e5eed6e7d2848|74693368|0|10.16.33.2|com_ops_dragonfly|1151.427|
[2019-01-07 14:06:32,780] INFO sign:345064-1546868841.351 lineno:111 : download FAIL cost:1151.429s length:74693368 reason:0
```
看supernode日志/home/admin/supernode/logs/app.log看不出什么问题，感觉supernode是下载成功的：
```
2019-01-07 13:47:23.759  INFO 9 --- [http-nio-8080-exec-1] c.a.d.s.repository.TaskRepository        : get file length:74693368 from http client about taskId:c676b9e09f098a37d06e5fb55a2d42b8abf735f00a1cb5408c7072d36455720c
2019-01-07 13:47:23.759  INFO 9 --- [http-nio-8080-exec-1] c.a.d.s.service.impl.CdnManagerImpl      : do trigger cdn start for taskId:c676b9e09f098a37d06e5fb55a2d42b8abf735f00a1cb5408c7072d36455720c,httpLen:74693368
2019-01-07 13:47:23.760  INFO 9 --- [Thread-7] c.a.d.s.service.impl.CdnManagerImpl      : do trigger cdn success for taskId:c676b9e09f098a37d06e5fb55a2d42b8abf735f00a1cb5408c7072d36455720c
2019-01-07 13:47:23.763  INFO 9 --- [pool-1-thread-2] c.a.d.supernode.service.cdn.Downloader   : taskId:c676b9e09f098a37d06e5fb55a2d42b8abf735f00a1cb5408c7072d36455720c fileUrl:https://10.250.250.16/v2/devops/jry-java/blobs/sha256:0ffa5ac9f3c5c761212037c66effa9c800973aca2f0583dcd24e5eed6e7d2848 on downloader
2019-01-07 13:47:48.426  INFO 9 --- [Thread-8] c.a.d.s.service.impl.CdnReporterImpl     : taskId:c676b9e09f098a37d06e5fb55a2d42b8abf735f00a1cb5408c7072d36455720c fileLength:74693458 status:SUCCESS from:local
2019-01-07 13:47:48.426  INFO 9 --- [Thread-8] c.a.d.supernode.service.cdn.SuperWriter  : taskId:c676b9e09f098a37d06e5fb55a2d42b8abf735f00a1cb5408c7072d36455720c readCost:24004,totalCost:24663,fileLength:74693368,realMd5:84ff00a39874c6f4d68bc23c74575b0f
```
想了解
1.卡死的原因是什么？
2.如果能判断出造成卡死的原因是否能不要等待这么长时间，直接重试

### LINKED ISSUE #545: After deployment, it can be used normally, but after a period of time, it is impossible to retrieve the image.
Using k8s deployment, deployment deployment deployment supernode, daemonset deployment dfclient

// 使用k8s部署，deployment部署supernode，daemonset部署dfclient 

Configure httpproxy to pull the private image warehouse image.

// 配置httpproxy，拉取私有镜像仓库镜像。 

After deployment, the node can pull the mirror smoothly. Walk on the P2P network.

// 部署完成后，节点可以顺利拉取镜像。走的p2p网络。 

But after a day or two of use, there will be an inability to pull the mirror. The phenomenon is that when the mirror is pulled, part of the mirror layer is retry, and the final report is 502.

// 但是在使用一两天后，就会出现无法拉取镜像。现象就是拉镜像的时候，部分镜像层一直retry，最后报502. 

[root@paasn4 ~]# docker pull registry.paas/library/mysql:5.6
5.6: Pulling from library/mysql
6ae821421a7d: Downloading
a9e976e3aa6d: Download complete
e3735e44a020: Download complete
bfd564e9483f: Download complete
df705f26e488: Download complete
25097d3ea7b8: Download complete
606c1b74fda2: Download complete
eda5767f8d3f: Download complete
5b57c9a598ba: Download complete
57f28f294e2f: Download complete
94b29ab4a8ab: Download complete
received unexpected HTTP status: 502 Bad Gateway
The solution is to restart the pod of supernode at this time, and it will return to normal, but after using it for a day or two, the problem will arise again.

// 解决方法，此时重启supernode的pod，就会恢复正常，但是使用一两天后，又会出现该问题。 

Log information, supernode log no error, dfclient log error

// 日志信息，supernode日志无报错，dfclient日志有报错 

image

***!!!!WE STRONGLY ENCOURAGE YOU TO DESCRIBE YOUR ISSUE IN ENGLISH!!!!***

### PR DESCRIPTION
**Pull request by lowzj**
Signed-off-by: lowzj <zj3142063@gmail.com>

### Ⅰ. Describe what this PR did

dfget reports downloading errors to supernode, supernode checks and repairs local data automatically, currently repair these client errors:
* `FILE_NOT_EXIST`
* `FILE_MD5_NOT_MATCH`

add a new api into supernode to delete the task's data manually if any error occurred

```
DELETE /peer/tasks/{taskId}
```

client-side pr: #583 

> merge into branch `master` and `0.3.x`

### Ⅱ. Does this pull request fix one issue?

fixes #545 
fixes #311 
fixes #68

### CODE REVIEW

File: src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/service/repair/BaseClientErrorHandler.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                } catch (InterruptedException e) {
+                    interrupted = true;
+                    break;
+                }
+            }
+        }
+
+        public void add(String taskId, boolean last) {
+            if (interrupted) {
+                running.remove(taskId);
```
Comment (chenchaobing) [inline src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/service/repair/BaseClientErrorHandler.java]: remove and return ?

File: src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/service/repair/BaseClientErrorHandler.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            }
+            try {
+                if (last) {
+                    queue.addLast(taskId);
+                } else {
+                    queue.addFirst(taskId);
+                }
+            } catch (Throwable t) {
```
Comment (chenchaobing) [inline src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/service/repair/BaseClientErrorHandler.java]: should log the exception?

File: src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/service/repair/BaseClientErrorHandler.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+ * The client error will be handled immediately. In the process of handling,
+ * and within 5 seconds after completing handling, the subsequent client errors
+ * with the same taskId and error type will be ignored.
+ *
+ * @author lowzj
+ */
+@Slf4j
+public abstract class BaseClientErrorHandler implements ClientErrorHandler {
+    private static final ExecutorService executor = ExecutorsUtil
+        .newThreadPool("client-error", 20, 200);
```
Comment (chenchaobing) [inline src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/service/repair/BaseClientErrorHandler.java]: i guess may not need so many threads.

File: src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/service/repair/FileMd5NotMatchHandler.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    @Override
+    void doHandle(final ClientErrorInfo info) {
+        boolean gc = false;
+        String md5FromMeta = getMd5FromMeta(info.getTaskId(), info.getRange());
+        String localMd5 = getMd5(info.getTaskId(), info.getRange());
+        if (StringUtils.isNotBlank(localMd5)) {
+            if (!StringUtils.equals(md5FromMeta, localMd5) // data in supernode is not right
+                && !StringUtils.equals(localMd5, info.getExpectedMd5())) {
```
Comment (chenchaobing) [inline src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/service/repair/FileMd5NotMatchHandler.java]: i think if data in supernode is not right then we will remove the task from the super.
is condition !StringUtils.equals(localMd5, info.getExpectedMd5()) necessary?

File: src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/service/repair/BaseClientErrorHandler.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    public boolean isErrorType(ClientErrorType type) {
+        return errorType().equals(type);
+    }
+
+    @Override
+    public void handle(final ClientErrorInfo info) {
+        if (isInvalidInfo(info)) {
+            return;
+        }
+        Integer pre = running.putIfAbsent(info.getTaskId(), currentSeconds());
```
Comment (chenchaobing) [inline src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/service/repair/BaseClientErrorHandler.java]: one task may have different error.
should we distinguish different error codes?

File: src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/service/repair/FileMd5NotMatchHandler.java (hunk 4 above)
Comment (lowzj) [reply src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/service/repair/FileMd5NotMatchHandler.java]: Before supernode completing downloading file from remote server, the meta data is not written into local disk, so the `md5FromMeta` may be null.

**Comment by chenchaobing** [comment]
LGTM.

**Pull request merged**